abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1996 @@
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+  - `i32 (size_t) body_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called for each chunk of HTTP response body received from upstream,
+even when the processing is paused.
+
+Response body chunk (of `body_size`) can be retrieved and/or modified
+using [`proxy_get_buffer_bytes`] and/or [`proxy_set_buffer_bytes`]
+with `buffer_id` set to `HTTP_RESPONSE_BODY`.
+
+`body_size` represents total available size of the body that can be
+retrieved, and its value will increment if the processing is paused
+and the body is buffered by the host and not forwarded downstream.
+
+Paused HTTP response can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`HTTP_RESPONSE`.
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `HTTP_RESPONSE_BODY` buffer downstream.
+- `PAUSE` to pause processing.
+
+
+#### `proxy_on_response_trailers`
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+  - `i32 (size_t) num_trailers`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called when HTTP response trailers are received from upstream.
+
+All `num_trailers` trailers can be retrieved and/or modified using
+[`proxy_get_header_map_pairs`] and/or [`proxy_set_header_map_pairs`]
+with `map_id` set to `HTTP_RESPONSE_TRAILERS`.
+
+Individual trailers can be retrieved and/or modified using
+[`proxy_get_header_map_value`], [`proxy_replace_header_map_value`],
+[`proxy_add_header_map_value`] and/or [`proxy_remove_header_map_value`]
+with `map_id` set to `HTTP_RESPONSE_TRAILERS`.
+
+Paused request can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`HTTP_REQUEST`.
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `HTTP_RESPONSE_TRAILERS` fields downstream.
+- `PAUSE` to pause processing.
+
+
+### Functions exposed by the host
+
+#### `proxy_send_local_response`
+
+* params:
+  - `i32 (uint32_t) status_code`
+  - `i32 (const char *) status_code_details_data`
+  - `i32 (size_t) status_code_details_size`
+  - `i32 (const uint8_t *) body_data`
+  - `i32 (size_t) body_size`
+  - `i32 (const uint8_t *) serialized_headers_data`
+  - `i32 (size_t) serialized_headers_size`
+  - `i32 (uint32_t) grpc_status`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Sends HTTP response with body (`body_data`, `body_size`) and
+[serialized] headers (`serialized_headers_data`,
+`serialized_headers_size`).
+
+This can be used as long as HTTP response headers were not send
+downstream.
+
+Returned `status` value is:
+- `OK` on success.
+- `INVALID_MEMORY_ACCESS` when `status_code_details_data`,
+  `status_code_details_size`, `body_data`, `body_size`, `headers_data`
+  and/or `headers_size` point to invalid memory address.
+
+
+## HTTP calls
+
+### Functions exposed by the host
+
+#### `proxy_http_call`
```

Comment:
@PiotrSikora  Is the function `proxy_http_call` the replacement of `proxy_dispatch_http_call` in the [old spec](https://github.com/proxy-wasm/spec/commit/86719f51b29cc1fcdf2d07ae80a6cc7bae0bc99d?short_path=c4e3008#diff-c4e3008315244482bb4516fa4106c38d1028b05c508af70b1bea7a75b4b235e5L653) ?